apps/designer/app/canvas/canvas.tsx

```diff
@@ -56,6 +53,101 @@ import {
   config,
 } from "~/canvas/features/wrapper-component/text-editor";
 
+// TODO: move this somewhere else
+import { useDrag } from "~/shared/design-system/components/primitives/dnd/use-drag";
+import { useDropTarget } from "~/shared/design-system/components/primitives/dnd/use-drop-target";
+import { findInstanceById, getInstancePath } from "~/shared/tree-utils";
+import { primitives } from "~/shared/canvas-components";
+const useDragAndDrop = () => {
+  const [rootInstance] = useRootInstance();
+
+  const dropTargetHandlers = useDropTarget({
+    isDropTarget(element) {
+      return (
```

Comment:
why not doing the canAcceptChild check here instead of onDropTargetChange?

Reply:
We'll still need to be able to swap the target based on the area. I want to do all the swapping in one place. And also want to rely more on the data tree instead of DOM tree when choosing the target. I've built a new `useDropTarget` in the primitive that will make this easier. Will try to use it today here.

Reply:
right, forgot